#' @title Visualizing CytoTRACE results
#'
#' @description This function generates 2D plots and tables to visualize CytoTRACE, known phenotypes, and gene expression.
#' The current implementation uses t-SNE for dimensional reduction but users can also input their own embeddings.
#' At minimum, the ‘plotCytoTRACE’ function takes as input a list object generated by either the ‘CytoTRACE’ or ‘iCytoTRACE’ functions.
#' Users can also optionally provide phenotype labels or gene names to generate additional plots.
#' Boxplots of CytoTRACE by phenotype labels are automatically generated when phenotype labels are provided.
#'
#'cyto_obj = NULL, phenotype = NULL, gene = NULL, colors = NULL, emb = NULL
#' @param cyto_obj a list object generated by the 'CytoTRACE' or 'iCytoTRACE' functions
#' @param phenotype (optional) a character vector of phenotype labels corresponding to the single cells in 'cyto_obj'. The vector names should match the column names of the original gene expression matrix.
#' @param gene  (optional) a case-insensitive string corresponding to a gene of interest. The case-insensitive string should match values provided as row names of the original gene expression matrix.
#' @param colors  (optional) a character vector of colors to be used to create a palette for the plots
#' @param emb  (optional) a data frame containing two columns corresponding to the 2D embeddings and where the row names match the column names of the original gene expression matrix.
#' @param otherValue (optional) a numeric vector of values to plot with CytoTRACE. This will replace the 'gene' plot.
#' @param otherName (optional) a string to replace 'gene' or describe the input values in 'otherValue'
#' @param outputDir path to output directory where plots and tables will be written. Format should be "/PATH/TO/DIRECTORY/".
#'
#' @return a PDF of 2D embedded plots colored by CytoTRACE, and, if provided, phenotype labels, and gene expression.
#' @return a PDF of a boxplot of CytoTRACE by phenotype labels (if provided).
#' @return a tab-delimited text file containing a table of CytoTRACE values t-SNE embeddings, and, if provided, phenotype labels and gene expression values.
#'
#' @author Gunsagar Gulati <cytotrace@gmail.com>
#'
#' @seealso https://cytotrace.stanford.edu
#'
#' @references https://doi.org/10.1101/649848
#'
#' @examples
#'
#' #Use the bone marrow 10x scRNA-seq dataset to run CytoTRACE
#' results <- CytoTRACE(marrow_10x_expr)
#'
#' #Run plotCytoTRACE
#' plotCytoTRACE(results, phenotype = marrow_10x_pheno, gene = "Kit")
#'
#' @export

plotCytoTRACE <- function(cyto_obj = NULL, phenotype = NULL, gene = NULL, colors = NULL, emb = NULL, otherName = NULL, otherValue = NULL, outputDir = "./"){

mat <- cyto_obj$exprMatrix
cyto <- cyto_obj$CytoTRACE

##Setup phenotype labels
if(!is.null(phenotype)){
  if (ncol(mat) != length(phenotype)) {
    stop("The number of phenotype labels provided does not match the number of cells in the CytoTRACE object.", type="error")
  }
  if (length(intersect(colnames(mat), names(phenotype))) != length(phenotype)) {
    stop("The names of the phenotype labels do no match the names in the CytoTRACE object.", type="error")
  }
  if (max(nchar(phenotype))>25) {
    warning("Phenotype labels exceed 25 characters. This may result in overcrowded text in some of the visualizations.")
  }
  pheno <- phenotype[colnames(mat)]
} else {
  pheno <- NULL
}

##Setup gene
if(!is.null(gene)){
  gene <- tolower(gene)
  lower_case_names <- tolower(rownames(mat))
  if (gene %in% lower_case_names) {
    gene_exp <- mat[lower_case_names %in% gene,]
    gene2 <- rownames(mat)[which(lower_case_names == gene)]
  } else {
    gene_exp <- integer(length(mat[1,]))
    gene2 <- gene
  }
} else {
  gene_exp <- NULL
}

#Other
if(!is.null(otherValue)){
gene_exp <- otherValue
if(length(gene_exp) != length(cyto)){
  stop("The length of the 'otherValue' object does not match the length of the CytoTRACE object.")
}
}
if(!is.null(otherName)){
  gene2 <- ""
}

#Remove NAs
mat <- mat[,!is.na(cyto)]
pheno <- pheno[!is.na(cyto)]
gene_exp <- gene_exp[!is.na(cyto)]
cyto <- cyto[!is.na(cyto)]

#Run tSNE
dotsne <- function(mat, perplexity){

  # --------------------------------------------------
  # get variable genes from normalized UMI counts
  # --------------------------------------------------
  # m: matrix normalized by UMI counts
  .get_variable_gene<-function(m) {

    df<-data.frame(mean=colMeans(m),cv=apply(m,2,sd)/colMeans(m),var=apply(m,2,var))
    df$dispersion<-with(df,var/mean)
    df$mean_bin<-with(df,cut(mean,breaks=c(-Inf,quantile(mean,seq(0.1,1,0.05)),Inf)))
    var_by_bin<- plyr::ddply(df,"mean_bin",function(x) {
      data.frame(bin_median=median(x$dispersion),
                 bin_mad=mad(x$dispersion))
    })
    df$bin_disp_median<-var_by_bin$bin_median[match(df$mean_bin,var_by_bin$mean_bin)]
    df$bin_disp_mad<-var_by_bin$bin_mad[match(df$mean_bin,var_by_bin$mean_bin)]
    df$dispersion_norm<-with(df,abs(dispersion-bin_disp_median)/bin_disp_mad)
    df
  }
  # -----------------------------------------------------
  # compute top n principle components with propack
  # -----------------------------------------------------
  .do_propack <- function(x,n) {
    use_genes <- which(colSums(x) > 1)
    m <- x[,use_genes]
    bc_tot <- rowSums(m)
    median_tot <- median(bc_tot)
    m <- sweep(m, 1, median_tot/bc_tot, '*')
    m <- sweep(m, 2, colMeans(m), '-')
    m <- sweep(m, 2, apply(m, 2, sd), '/')
    ppk<-propack.svd(as.matrix(m),neig=n)
    pca<-t(ppk$d*t(ppk$u))
    list(ppk=ppk,pca=pca, m=m,use_genes=use_genes)
  }
  # ---------------------------------------------
  # normalize the gene barcode matrix by umi
  # ---------------------------------------------
  .normalize_by_umi <-function(x) {
    cs <- colSums(x)
    x_use_genes <- which(cs >= 1)
    x_filt<-x[,x_use_genes]
    rs<-rowSums(x_filt)
    rs_med<-median(rs)
    x_norm<-x_filt/(rs/rs_med + 1e-15)
    list(m=x_norm,use_genes=x_use_genes)
  }

  matt <- t(mat)
  m_n<-matt
  l<-.normalize_by_umi(matt)
  m_n<-l$m
  df<-.get_variable_gene(m_n)
  disp_cut_off<-sort(df$dispersion_norm,decreasing=T)[1000]
  df$used<-df$dispersion_norm >= disp_cut_off
  set.seed(0)
  m_n_1000<-m_n[,head(order(-df$dispersion_norm),1000)]
  tsne <- Rtsne::Rtsne(m_n_1000, pca = T, check_duplicates = FALSE, perplexity = perplexity)
  tsn <- tsne$Y
  rownames(tsn) <- colnames(mat)
  return(tsn)
}

if(!is.null(emb)){
  if(length(intersect(colnames(mat), rownames(emb))) != ncol(mat)){
    stop("The samples in the embedding table do not match the samples in the CytoTRACE object.", type="error")
  }
  emb <- emb[colnames(mat),]
  tsne.proj <- emb
} else if (length(grep("coord", names(cyto_obj)))>0){
    emb <- cyto_obj$coord
    tsne.proj <- emb
  } else {
    message("Running t-SNE. To use your own coordinates, use the 'emb' flag.")
n_samples <- ncol(mat)
perplexity <- if(n_samples > 100) 30 else 10
tsne.proj <- data.frame(dotsne(mat, perplexity))
}

##Organize inputs
output_data <- data.frame(CytoTRACE=cyto,
                          Component1=tsne.proj[,1],
                          Component2=tsne.proj[,2])
if(!is.null(pheno)) output_data$Phenotype <- pheno
if(!is.null(gene_exp)) output_data$Gene <- gene_exp


#####
drawPlots <- function(data, colors, gene2) {
  data <- data[order(data$CytoTRACE),]
  emb <- cbind.data.frame(Component1 = data$Component1, Component2 = data$Component2)

  #Setup colors and other
  temp <- RColorBrewer::brewer.pal(11, "Spectral")
  temp[6] <- "gold"
  rbPal <- colorRampPalette(temp)

  if(!is.null(colors)){
    rbPal <- colorRampPalette(colors)
  }

  ###Adjust point size
  if(length(data$CytoTRACE) <= 200){
    ptsize <- 4
  } else if (length(data$CytoTRACE) > 200 & length(data$CytoTRACE) < 1000) {
    ptsize <- 3.5
  } else if (length(data$CytoTRACE) > 1000 & length(data$CytoTRACE) < 2000) {
    ptsize <- 3
  } else if (length(data$CytoTRACE) > 2000 & length(data$CytoTRACE) < 5000) {
    ptsize <- 2
  }else if (length(data$CytoTRACE) > 5000 & length(data$CytoTRACE) < 10000) {
    ptsize <- 1
  } else if (length(data$CytoTRACE) >= 10000) {
    ptsize <- 0.5
  }

  #######GRAPH##########
  #Phenotype labels
  if (!is.null(data$Phenotype)){
    data$Phenotype <- substring(data$Phenotype, 1, 25)
    suppressWarnings(aggregated <- aggregate(data, by=list(data$Phenotype), FUN=median))
    levels = as.character(aggregated[order(-aggregated$CytoTRACE),]$Group.1)
    data$Phenotype <- factor(data$Phenotype, levels=levels)
    pheno_levels <- levels(data$Phenotype)
    cols <- rbPal(length(pheno_levels))
    names(cols) <- pheno_levels
    pdf(paste0(outputDir, "CytoTRACE_by_Phenotype_Boxplot.pdf"), width = length(pheno_levels)*1.5, height = 6)
    par(mar = c(log(max(nchar(pheno_levels))+2, 2)*3.25,6,2,1), xpd = NA)
    boxplot(data$CytoTRACE~data$Phenotype, outline = F,las = 2,xaxt = "n", yaxt = "n",
            staplelwd = 2,
            medlwd = 2,
            whisklty = 1,
            border = cols,
            col = adjustcolor(cols, alpha.f = 0.25),
            ylab = "Predicted ordering by CytoTRACE",
            xlab = "",
            cex.lab = 1.75,
            frame.plot = F,
            xlim = c(0, length(pheno_levels)+0.5),
            ylim = c(min(data$CytoTRACE),1))
    mtext("Cell phenotypes",
          cex = 1.75, side = 1, line = round(max(nchar(pheno_levels))/2,1) + 1)
    axis(1, pos =0, at = 1:length(pheno_levels), labels = FALSE)
    text(x =  seq_along(pheno_levels), y = -0.05, srt = 45, adj = 1,
         labels = pheno_levels, xpd = TRUE,cex = 1.5)
    segments(0,0,length(pheno_levels)+0.5,0)
    axis(2, pos = 0, at = seq(0, 1, 0.2), labels = format(seq(0, 1, 0.2), nsmall = 1),las = 2, cex.axis = 1.5)
    segments(0, 0, 0, 1)
    stripchart(data$CytoTRACE~data$Phenotype, vertical = TRUE,
               method = "jitter", add = TRUE, pch = 16, col = cols, bg = 'white',
               lwd = 1, cex = 2/round(log(length(data$CytoTRACE), 10)))
    dev.off()

    suppressWarnings(aggregated <- aggregate(data, by=list(data$Phenotype), FUN=mean))
    levels = as.character(aggregated[order(-aggregated$CytoTRACE),]$Group.1)
    data$Phenotype <- factor(data$Phenotype, levels=levels)
    pheno_levels <- levels(data$Phenotype)
    cols <- rbPal(length(pheno_levels))
    names(cols) <- pheno_levels

    p1 <- ggplot2::ggplot(data, ggplot2::aes(x=emb[,1], y=emb[,2])) +
      ggplot2::geom_point(ggplot2::aes(colour = Phenotype), size = ptsize)+
      ggplot2::scale_colour_manual(values = cols,
                          guide = ggplot2::guide_legend(frame.colour = "black",
                                               override.aes = list(size = 4)))+
      ggplot2::labs(x = colnames(emb)[1], y = colnames(emb)[2], title = "Phenotype")+
      ggpubr::theme_pubr()+
      ggplot2::theme(
        legend.text = ggplot2::element_text(size = 14),
        legend.title = ggplot2::element_blank(),
        plot.title = ggplot2::element_text(size = 21, hjust = 0.5),
        axis.title.x = ggplot2::element_text(size = 18),
        axis.title.y = ggplot2::element_text(size = 18),
        axis.text=ggplot2::element_text(size=16),
        legend.position="right",
        plot.margin = ggplot2::unit(c(0.5,1,0.5,1), "cm")
      )

  }

  if (!is.null(data$Gene) & sum(data$Gene) != 0) {
    p2 <- ggplot2::ggplot(data, ggplot2::aes(x=emb[,1], y=emb[,2])) +
      ggplot2::geom_point(ggplot2::aes(colour = Gene), size = ptsize)+
      ggplot2::scale_colour_gradientn(name=ifelse(is.null(otherName), "Gene\nexpression", otherName),
                             colours = rev(rbPal(50)),
                             guide = ggplot2::guide_colourbar(ticks.colour = "black",
                                                     ticks.linewidth = 1,
                                                     frame.colour = "black"))+
      ggplot2::labs(x = colnames(emb)[1], y = colnames(emb)[2], title = bquote(italic(.(gene2))))+
      ggpubr::theme_pubr()+
      ggplot2::theme(
        legend.text = ggplot2::element_text(size = 12),
        legend.title = ggplot2::element_text(size = 14),
        plot.title = ggplot2::element_text(size = 21, hjust = 0.5),
        axis.title.x = ggplot2::element_text(size = 18),
        axis.title.y = ggplot2::element_text(size = 18),
        axis.text=ggplot2::element_text(size=16),
        legend.position="right",
        plot.margin = ggplot2::unit(c(0.5,1,0.5,1), "cm")
      )
  } else if (!is.null(data$Gene) & sum(data$Gene) == 0) {
    p2 <- ggplot2::ggplot(data, ggplot2::aes(x=emb[,1], y=emb[,2])) +
      ggplot2::geom_point(ggplot2::aes(x = emb[,1], y = emb[,2], colour = Gene), size = 0.1)+
      ggplot2::scale_colour_gradientn(name="Gene\nexpression",
                             colours = "white",
                             guide = ggplot2::guide_colourbar(ticks.colour = "black",
                                                     ticks.linewidth = 1,
                                                     frame.colour = "black"))+
      ggplot2::annotate("text", x = mean(emb[,1]), y = mean(emb[,2]), xmin = min(emb[,1]), xmax = max(emb[,1]),
               ymin = min(emb[,2]), ymax = max(emb[,2]),
               label = "Selected gene\nis not expressed in dataset", size = 6)+
      ggplot2::labs(x = colnames(emb)[1], y = colnames(emb)[2], title = bquote(italic(.(gene2))))+
      ggpubr::theme_pubr()+
      ggplot2::theme(
        legend.text = ggplot2::element_text(size = 12),
        legend.title = ggplot2::element_text(size = 14),
        plot.title = ggplot2::element_text(size = 21, hjust = 0.5),
        axis.title.x = ggplot2::element_text(size = 18),
        axis.title.y = ggplot2::element_text(size = 18),
        axis.text=ggplot2::element_text(size=16),
        legend.position="right",
        plot.margin = ggplot2::unit(c(0.5,1,0.5,1), "cm")
      )

  }

  #CytoTRACE
  p3 <- ggplot2::ggplot(data, ggplot2::aes(x=emb[,1], y=emb[,2])) +
    ggplot2::geom_point(ggplot2::aes(colour = CytoTRACE), size = ptsize)+
    ggplot2::scale_colour_gradientn(name = "Predicted\norder",
                           colours = rev(rbPal(50)),
                           guide = ggplot2::guide_colourbar(ticks.colour = "black",
                                                   ticks.linewidth = 1,
                                                   frame.colour = "black"),
                           breaks = seq(0, 1, 0.2),
                           labels=c("0.0 (More diff.)", 0.2,0.4, 0.6, 0.8, "1.0 (Less diff.)"))+
    ggplot2::labs(x = colnames(emb)[1], y = colnames(emb)[2], title = "CytoTRACE")+
    ggpubr::theme_pubr()+
    ggplot2::theme(
      legend.text = ggplot2::element_text(size = 12),
      legend.title = ggplot2::element_text(size = 14),
      plot.title = ggplot2::element_text(size = 21, hjust = 0.5),
      axis.title.x = ggplot2::element_text(size = 18),
      axis.title.y = ggplot2::element_text(size = 18),
      axis.text=ggplot2::element_text(size=16),
      legend.position="left",
      plot.margin = ggplot2::unit(c(0.5,1,0.5,1), "cm")

    )

  if (!is.null(data$Phenotype) & !is.null(data$Gene)){
    mp <- egg::ggarrange(p3, p1, p2, nrow = 1, newpage = T)
      ggpubr::ggexport(mp, filename = paste0(outputDir, "CytoTRACE_plot.pdf"),
                       width = 21, height= 5, verbose = F)
  } else if (!is.null(data$Phenotype) & is.null(data$Gene)) {
   mp <-  egg::ggarrange(p3, p1, nrow = 1)
   ggpubr::ggexport(mp, filename = paste0(outputDir, "CytoTRACE_plot.pdf"),
                    width = 14, height= 5, verbose = F)
  } else if (is.null(data$Phenotype) & !is.null(data$Gene)) {
    mp <- egg::ggarrange(p3, p2, nrow = 1)
    ggpubr::ggexport(mp, filename = paste0(outputDir, "CytoTRACE_plot.pdf"),
                     width = 14, height= 5, verbose = F)
  } else if (is.null(data$Phenotype) & is.null(data$Gene)) {
    mp <- egg::ggarrange(p3, nrow = 1)
    ggpubr::ggexport(mp, filename = paste0(outputDir, "CytoTRACE_plot.pdf"),
                     width = 7, height= 5, verbose = F)
  }

}
drawPlots(output_data, colors, gene2)
write.table(output_data, paste0(outputDir, "CytoTRACE_plot_table.txt"), sep = "\t", quote = F)
}



